# 1.Write a Pandas program to create a dataframe from a dictionary and display it.
Sample data: {'X':[78,85,96,80,86], 'Y':[84,94,89,83,86],'Z':[86,97,96,72,83]}
Expected Output:
    X   Y   Z                                                         
0  78  84  86                                                        
1  85  94  97                                                         
2  96  89  96                                                      
3  80  83  72                                                         
4  86  86  83

In [1]:
import pandas as pd

# Create the dictionary
data = {'X': [78, 85, 96, 80, 86],
        'Y': [84, 94, 89, 83, 86],
        'Z': [86, 97, 96, 72, 83]}

# Create the DataFrame
df = pd.DataFrame(data)

# Display the DataFrame
print(df)


    X   Y   Z
0  78  84  86
1  85  94  97
2  96  89  96
3  80  83  72
4  86  86  83


# 2. Write a Pandas program to create and display a DataFrame from a specified dictionary data which has the index labels.
Sample Python dictionary data and list labels:
exam_data = {'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']
Expected Output:
   attempts       name qualify  score                              
a         1  Anastasia     yes   12.5                                 
b         3       Dima      no    9.0                                 
....                              
i         2      Kevin      no    8.0                                
j         1      Jonas     yes   19.0

In [2]:
import pandas as pd
import numpy as np

# Define the dictionary data
exam_data = {
    'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
    'score': [12.5, 9, 16.5, np.nan, 9, 20, 14.5, np.nan, 8, 19],
    'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
    'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes']
}
labels = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j']

# Create the DataFrame with index labels
df = pd.DataFrame(exam_data, index=labels)

# Display the DataFrame
print(df)


        name  score  attempts qualify
a  Anastasia   12.5         1     yes
b       Dima    9.0         3      no
c  Katherine   16.5         2     yes
d      James    NaN         3      no
e      Emily    9.0         2      no
f    Michael   20.0         3     yes
g    Matthew   14.5         1     yes
h      Laura    NaN         1      no
i      Kevin    8.0         2      no
j      Jonas   19.0         1     yes


# 47. Write a Pandas program to get the specified row value of a given DataFrame.
Sample data:
Original DataFrame
   col1  col2  col3
0     1     4     7
1     2     5     8
2     3     6    12
3     4     9     1
4     7     5    11
Value of Row
col1    1
col2    4
col3    7
Name: 0, dtype: int64
Value of Row4
col1    4
col2    9
col3    1
Name: 3, dtype: int64

In [3]:
import pandas as pd

# Create the original DataFrame
data = {'col1': [1, 2, 3, 4, 7],
        'col2': [4, 5, 6, 9, 5],
        'col3': [7, 8, 12, 1, 11]}

df = pd.DataFrame(data)

# Print the original DataFrame
print("Original DataFrame:")
print(df)

# Get the value of Row 0
row_value_0 = df.loc[0]

# Print the value of Row 0
print("\nValue of Row 0:")
print(row_value_0)

# Get the value of Row 3
row_value_3 = df.loc[3]

# Print the value of Row 3
print("\nValue of Row 3:")
print(row_value_3)


Original DataFrame:
   col1  col2  col3
0     1     4     7
1     2     5     8
2     3     6    12
3     4     9     1
4     7     5    11

Value of Row 0:
col1    1
col2    4
col3    7
Name: 0, dtype: int64

Value of Row 3:
col1    4
col2    9
col3    1
Name: 3, dtype: int64


# 48. Write a Pandas program to get the datatypes of columns of a DataFrame.
Sample data:
Original DataFrame:
   attempts       name qualify  score
0         1  Anastasia     yes   12.5
1         3       Dima      no    9.0
.......
8         2      Kevin      no    8.0
9         1      Jonas     yes   19.0
Data types of the columns of the said DataFrame:
attempts      int64
name         object
qualify      object
score       float64
dtype: object

In [4]:
import pandas as pd

# Create the original DataFrame
data = {'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
        'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
        'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes'],
        'score': [12.5, 9.0, 16.5, 10.0, 9.0, 20.0, 14.5, None, 8.0, 19.0]}

df = pd.DataFrame(data)

# Print the original DataFrame
print("Original DataFrame:")
print(df)

# Get the datatypes of the columns
column_datatypes = df.dtypes

# Print the datatypes
print("\nData types of the columns of the said DataFrame:")
print(column_datatypes)


Original DataFrame:
   attempts       name qualify  score
0         1  Anastasia     yes   12.5
1         3       Dima      no    9.0
2         2  Katherine     yes   16.5
3         3      James      no   10.0
4         2      Emily      no    9.0
5         3    Michael     yes   20.0
6         1    Matthew     yes   14.5
7         1      Laura      no    NaN
8         2      Kevin      no    8.0
9         1      Jonas     yes   19.0

Data types of the columns of the said DataFrame:
attempts      int64
name         object
qualify      object
score       float64
dtype: object


# 49. Write a Pandas program to append data to an empty DataFrame.
Sample data:
Original DataFrame:
After appending some data:
   col1  col2
0     0     0
1     1     1
2     2     2

In [5]:
import pandas as pd

# Create an empty DataFrame
df = pd.DataFrame()

# Print the original empty DataFrame
print("Original DataFrame:")
print(df)

# Append some data to the DataFrame
data = {'col1': [0, 1, 2], 'col2': [0, 1, 2]}
df = df.append(pd.DataFrame(data))

# Print the DataFrame after appending data
print("\nAfter appending some data:")
print(df)


Original DataFrame:
Empty DataFrame
Columns: []
Index: []

After appending some data:
   col1  col2
0     0     0
1     1     1
2     2     2


C:\Users\Akshit Koche\AppData\Local\Temp\ipykernel_16720\2957597319.py:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append(pd.DataFrame(data))


# 50. Write a Pandas program to sort a given DataFrame by two or more columns.
Sample data:
Original DataFrame:
   attempts       name qualify  score
0         1  Anastasia     yes   12.5 
1         3       Dima      no    9.0
........
8         2      Kevin      no    8.0
9         1      Jonas     yes   19.0

Sort the above DataFrame on attempts, name:
   attempts       name qualify  score
0         1  Anastasia     yes   12.5
9         1      Jonas     yes   19.0
7         1      Laura      no    NaN
6         1    Matthew     yes   14.5
4         2      Emily      no    9.0
2         2  Katherine     yes   16.5
8         2      Kevin      no    8.0
1         3       Dima      no    9.0
3         3      James      no    NaN
5         3    Michael     yes   20.0

In [6]:
import pandas as pd

# Create the original DataFrame
df = pd.DataFrame({
    'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
    'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
    'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes'],
    'score': [12.5, 9.0, 16.5, None, 9.0, 20.0, 14.5, None, 8.0, 19.0]
})

# Print the original DataFrame
print("Original DataFrame:")
print(df)

# Sort the DataFrame on attempts, name
sorted_df = df.sort_values(['attempts', 'name'])

# Print the sorted DataFrame
print("\nSorted DataFrame:")
print(sorted_df)


Original DataFrame:
   attempts       name qualify  score
0         1  Anastasia     yes   12.5
1         3       Dima      no    9.0
2         2  Katherine     yes   16.5
3         3      James      no    NaN
4         2      Emily      no    9.0
5         3    Michael     yes   20.0
6         1    Matthew     yes   14.5
7         1      Laura      no    NaN
8         2      Kevin      no    8.0
9         1      Jonas     yes   19.0

Sorted DataFrame:
   attempts       name qualify  score
0         1  Anastasia     yes   12.5
9         1      Jonas     yes   19.0
7         1      Laura      no    NaN
6         1    Matthew     yes   14.5
4         2      Emily      no    9.0
2         2  Katherine     yes   16.5
8         2      Kevin      no    8.0
1         3       Dima      no    9.0
3         3      James      no    NaN
5         3    Michael     yes   20.0


# 51. Write a Pandas program to convert the datatype of a given column (floats to ints).
Sample data:
Original DataFrame:
attempts name qualify score
0 1 Anastasia yes 12.50
1 3 Dima no 9.10
......
8 2 Kevin no 8.80
9 1 Jonas yes 19.13
Data types of the columns of the said DataFrame:
attempts int64
name object
qualify object
score float64
dtype: object
Now change the Data type of 'score' column from float to int:
attempts name qualify score
0 1 Anastasia yes 12
1 3 Dima no 9
2 2 Katherine yes 16
3 3 James no 12
4 2 Emily no 9
5 3 Michael yes 20
6 1 Matthew yes 14
7 1 Laura no 11
8 2 Kevin no 8
9 1 Jonas yes 19
Data types of the columns of the DataFrame now:
attempts int64
name object
qualify object
score int64
dtype: object

In [7]:
import pandas as pd

# Create the original DataFrame
df = pd.DataFrame({
    'attempts': [1, 3, 2, 3, 2, 3, 1, 1, 2, 1],
    'name': ['Anastasia', 'Dima', 'Katherine', 'James', 'Emily', 'Michael', 'Matthew', 'Laura', 'Kevin', 'Jonas'],
    'qualify': ['yes', 'no', 'yes', 'no', 'no', 'yes', 'yes', 'no', 'no', 'yes'],
    'score': [12.50, 9.10, 16.29, 12.40, 9.35, 20.00, 14.50, 11.20, 8.80, 19.13]
})

# Print the original DataFrame
print("Original DataFrame:")
print(df)

# Check the data types of the columns
print("Data types of the columns of the said DataFrame:")
print(df.dtypes)

# Convert the datatype of 'score' column from float to int
df['score'] = df['score'].astype(int)

# Print the DataFrame after changing the data type
print("DataFrame after changing the data type of 'score' column:")
print(df)

# Check the data types of the columns again
print("Data types of the columns of the DataFrame now:")
print(df.dtypes)


Original DataFrame:
   attempts       name qualify  score
0         1  Anastasia     yes  12.50
1         3       Dima      no   9.10
2         2  Katherine     yes  16.29
3         3      James      no  12.40
4         2      Emily      no   9.35
5         3    Michael     yes  20.00
6         1    Matthew     yes  14.50
7         1      Laura      no  11.20
8         2      Kevin      no   8.80
9         1      Jonas     yes  19.13
Data types of the columns of the said DataFrame:
attempts      int64
name         object
qualify      object
score       float64
dtype: object
DataFrame after changing the data type of 'score' column:
   attempts       name qualify  score
0         1  Anastasia     yes     12
1         3       Dima      no      9
2         2  Katherine     yes     16
3         3      James      no     12
4         2      Emily      no      9
5         3    Michael     yes     20
6         1    Matthew     yes     14
7         1      Laura      no     11
8         2      Ke

# 52. Write a Pandas program to remove infinite values from a given DataFrame.
Sample data:
Original DataFrame:
0
0 1000.000000
1 2000.000000
2 3000.000000
3 -4000.000000
4 inf
5 -inf
Removing infinite values:
0
0 1000.0
1 2000.0
2 3000.0
3 -4000.0
4 NaN
5 NaN

# 53. Write a Pandas program to insert a given column at a specific column index in a DataFrame.
Sample data:
Original DataFrame
col2 col3
0 4 7
1 5 8
2 6 12
3 9 1
4 5 11
New DataFrame
col1 col2 col3
0 1 4 7
1 2 5 8
2 3 6 12
3 4 9 1
4 7 5 11

# 54. Write a Pandas program to convert a given list of lists into a Dataframe.
Sample data:
Original list of lists:
[[2, 4], [1, 3]]
New DataFrame
col1 col2
0 2 4
1 1 3

In [8]:
import pandas as pd

# Original list of lists
original_list = [[2, 4], [1, 3]]

# Convert list of lists to DataFrame
df = pd.DataFrame(original_list, columns=['col1', 'col2'])

# Print the new DataFrame
print("New DataFrame:")
print(df)


New DataFrame:
   col1  col2
0     2     4
1     1     3


# 55. Write a Pandas program to group by the first column and get second column as lists in rows.
Sample data:
Original DataFrame
col1 col2
0 C1 1
1 C1 2
2 C2 3
3 C2 3
4 C2 4
5 C3 6
6 C2 5
Group on the col1:
col1
C1 [1, 2]
C2 [3, 3, 4, 5]
C3 [6]
Name: col2, dtype: object

In [9]:
import pandas as pd

# Original DataFrame
df = pd.DataFrame({'col1': ['C1', 'C1', 'C2', 'C2', 'C2', 'C3', 'C2'],
                   'col2': [1, 2, 3, 3, 4, 6, 5]})

# Group by col1 and get col2 as lists in rows
result = df.groupby('col1')['col2'].apply(list)

# Print the result
print(result)


col1
C1          [1, 2]
C2    [3, 3, 4, 5]
C3             [6]
Name: col2, dtype: object


# 56. Write a Pandas program to get column index from column name of a given DataFrame.
Sample Output:
Original DataFrame
col1 col2 col3
0 1 4 7
1 2 5 8
2 3 6 12
3 4 9 1
4 7 5 11
Index of 'col2'
1

In [10]:
import pandas as pd

# Create the original DataFrame
data = {'col1': [1, 2, 3, 4, 7],
        'col2': [4, 5, 6, 9, 5],
        'col3': [7, 8, 12, 1, 11]}
df = pd.DataFrame(data)

# Get the column index of 'col2'
column_name = 'col2'
column_index = df.columns.get_loc(column_name)

# Print the column index
print("Index of '{}': {}".format(column_name, column_index))


Index of 'col2': 1


# 57. Write a Pandas program to count number of columns of a DataFrame.
Sample Output:
Original DataFrame
col1 col2 col3
0 1 4 7
1 2 5 8
2 3 6 12
3 4 9 1
4 7 5 11
Number of columns:
3

In [11]:
import pandas as pd

# Create the original DataFrame
data = {'col1': [1, 2, 3, 4, 7],
        'col2': [4, 5, 6, 9, 5],
        'col3': [7, 8, 12, 1, 11]}
df = pd.DataFrame(data)

# Count the number of columns
num_columns = df.shape[1]

# Print the number of columns
print("Number of columns:", num_columns)


Number of columns: 3


# 58. Write a Pandas program to select all columns, except one given column in a DataFrame.
Sample Output:
Original DataFrame
col1 col2 col3
0 1 4 7
1 2 5 8
2 3 6 12
3 4 9 1
4 7 5 11
All columns except 'col3':
col1 col2
0 1 4
1 2 5
2 3 6
3 4 9
4 7 5

In [12]:
import pandas as pd

# Create the original DataFrame
df = pd.DataFrame({'col1': [1, 2, 3, 4, 7],
                   'col2': [4, 5, 6, 9, 5],
                   'col3': [7, 8, 12, 1, 11]})

# Print the original DataFrame
print("Original DataFrame:")
print(df)

# Select all columns except 'col3'
columns_except_col3 = df.drop('col3', axis=1)
print("All columns except 'col3':")
print(columns_except_col3)


Original DataFrame:
   col1  col2  col3
0     1     4     7
1     2     5     8
2     3     6    12
3     4     9     1
4     7     5    11
All columns except 'col3':
   col1  col2
0     1     4
1     2     5
2     3     6
3     4     9
4     7     5


# 59. Write a Pandas program to get first n records of a DataFrame.
Sample Output:
Original DataFrame
col1 col2 col3
0 1 4 7
1 2 5 5
2 3 6 8
3 4 9 12
4 7 5 1
5 11 0 11
First 3 rows of the said DataFrame':
col1 col2 col3
0 1 4 7
1 2 5 5
2 3 6 8

In [13]:
import pandas as pd

# Create the original DataFrame
df = pd.DataFrame({'col1': [1, 2, 3, 4, 7, 11],
                   'col2': [4, 5, 6, 9, 5, 0],
                   'col3': [7, 5, 8, 12, 1, 11]})

# Print the original DataFrame
print("Original DataFrame:")
print(df)

# Get the first n rows of the DataFrame
n = 3  # Number of rows to get
first_n_rows = df.head(n)
print("First {} rows of the DataFrame:".format(n))
print(first_n_rows)


Original DataFrame:
   col1  col2  col3
0     1     4     7
1     2     5     5
2     3     6     8
3     4     9    12
4     7     5     1
5    11     0    11
First 3 rows of the DataFrame:
   col1  col2  col3
0     1     4     7
1     2     5     5
2     3     6     8


# 60. Write a Pandas program to get last n records of a DataFrame.
Sample Output:
Original DataFrame
col1 col2 col3
0 1 4 7
1 2 5 5
2 3 6 8
3 4 9 12
4 7 5 1
5 11 0 11
Last 3 rows of the said DataFrame':
col1 col2 col3
3 4 9 12
4 7 5 1
5 11 0 11

In [14]:
import pandas as pd

# Original DataFrame
df = pd.DataFrame({'col1': [1, 2, 3, 4, 7, 11],
                   'col2': [4, 5, 6, 9, 5, 0],
                   'col3': [7, 5, 8, 12, 1, 11]})

# Get the last 3 rows of the DataFrame
last_n_rows = df.tail(3)

# Print the last n rows
print("Last 3 rows of the DataFrame:")
print(last_n_rows)


Last 3 rows of the DataFrame:
   col1  col2  col3
3     4     9    12
4     7     5     1
5    11     0    11


# 61. Write a Pandas program to get topmost n records within each group of a DataFrame.
Sample Output:
Original DataFrame
col1 col2 col3
0 1 4 7
1 2 5 5
2 3 6 8
3 4 9 12
4 7 5 1
5 11 0 11
topmost n records within each group of a DataFrame:
col1 col2 col3
5 11 0 11
4 7 5 1
3 4 9 12
col1 col2 col3
3 4 9 12
2 3 6 8
1 2 5 5
4 7 5 1
col1 col2 col3
3 4 9 12
5 11 0 11
2 3 6 8

In [15]:
import pandas as pd

# Create the original DataFrame
df = pd.DataFrame({'col1': [1, 2, 3, 4, 7, 11],
                   'col2': [4, 5, 6, 9, 5, 0],
                   'col3': [7, 5, 8, 12, 1, 11]})

# Print the original DataFrame
print("Original DataFrame:")
print(df)

# Get topmost n records within each group
n = 3  # Number of records to get from each group
grouped = df.groupby('col1').head(n)
print("Topmost {} records within each group of a DataFrame:".format(n))
print(grouped)


Original DataFrame:
   col1  col2  col3
0     1     4     7
1     2     5     5
2     3     6     8
3     4     9    12
4     7     5     1
5    11     0    11
Topmost 3 records within each group of a DataFrame:
   col1  col2  col3
0     1     4     7
1     2     5     5
2     3     6     8
3     4     9    12
4     7     5     1
5    11     0    11


# 62. Write a Pandas program to remove first n rows of a given DataFrame.
Sample Output:
Original DataFrame
col1 col2 col3
0 1 4 7
1 2 5 5
2 3 6 8
3 4 9 12
4 7 5 1
5 11 0 11
After removing first 3 rows of the said DataFrame:
col1 col2 col3
3 4 9 12
4 7 5 1
5 11 0 11

In [16]:
import pandas as pd

# Create the original DataFrame
data = {'col1': [1, 2, 3, 4, 7, 11],
        'col2': [4, 5, 6, 9, 5, 0],
        'col3': [7, 5, 8, 12, 1, 11]}
df = pd.DataFrame(data)

# Remove the first 3 rows
n = 3
df = df.iloc[n:]

# Print the updated DataFrame
print(df)


   col1  col2  col3
3     4     9    12
4     7     5     1
5    11     0    11


# 63. Write a Pandas program to remove last n rows of a given DataFrame.
Sample Output:
Original DataFrame
col1 col2 col3
0 1 4 7
1 2 5 5
2 3 6 8
3 4 9 12
4 7 5 1
5 11 0 11
After removing last 3 rows of the said DataFrame:
col1 col2 col3
0 1 4 7
1 2 5 5
2 3 6 8

In [17]:
import pandas as pd

# Create the original DataFrame
data = {'col1': [1, 2, 3, 4, 7, 11],
        'col2': [4, 5, 6, 9, 5, 0],
        'col3': [7, 5, 8, 12, 1, 11]}
df = pd.DataFrame(data)

# Remove the last 3 rows
n = 3
df = df.iloc[:-n]

# Print the updated DataFrame
print(df)


   col1  col2  col3
0     1     4     7
1     2     5     5
2     3     6     8


# 64. Write a Pandas program to add a prefix or suffix to all columns of a given DataFrame.
Sample Output:
Original DataFrame
W X Y Z
0 68 78 84 86
1 75 85 94 97
2 86 96 89 96
3 80 80 83 72
4 66 86 86 83
Add prefix:
A_W A_X A_Y A_Z
0 68 78 84 86
1 75 85 94 97
2 86 96 89 96
3 80 80 83 72
4 66 86 86 83
Add suffix:
W_1 X_1 Y_1 Z_1
0 68 78 84 86
1 75 85 94 97
2 86 96 89 96
3 80 80 83 72
4 66 86 86 83

In [18]:
import pandas as pd

# Create the original DataFrame
df = pd.DataFrame({'W': [68, 75, 86, 80, 66],
                   'X': [78, 85, 96, 80, 86],
                   'Y': [84, 94, 89, 83, 86],
                   'Z': [86, 97, 96, 72, 83]})

# Print the original DataFrame
print("Original DataFrame:")
print(df)

# Add prefix
prefixed_columns = df.add_prefix('A_')
print("Add prefix:")
print(prefixed_columns)

# Add suffix
suffixed_columns = df.add_suffix('_1')
print("Add suffix:")
print(suffixed_columns)


Original DataFrame:
    W   X   Y   Z
0  68  78  84  86
1  75  85  94  97
2  86  96  89  96
3  80  80  83  72
4  66  86  86  83
Add prefix:
   A_W  A_X  A_Y  A_Z
0   68   78   84   86
1   75   85   94   97
2   86   96   89   96
3   80   80   83   72
4   66   86   86   83
Add suffix:
   W_1  X_1  Y_1  Z_1
0   68   78   84   86
1   75   85   94   97
2   86   96   89   96
3   80   80   83   72
4   66   86   86   83


# 65. Write a Pandas program to reverse order (rows, columns) of a given DataFrame.
Sample Output:
Original DataFrame
W X Y Z
0 68 78 84 86
1 75 85 94 97
2 86 96 89 96
3 80 80 83 72
4 66 86 86 83
Reverse column order:
Z Y X W
0 86 84 78 68
1 97 94 85 75
2 96 89 96 86
3 72 83 80 80
4 83 86 86 66
Reverse row order:
W X Y Z
4 66 86 86 83
3 80 80 83 72
2 86 96 89 96
1 75 85 94 97
0 68 78 84 86
Reverse row order and reset index:
W X Y Z
0 66 86 86 83
1 80 80 83 72
2 86 96 89 96
3 75 85 94 97
4 68 78 84 86

In [19]:
import pandas as pd

# Create the original DataFrame
df = pd.DataFrame({'W': [68, 75, 86, 80, 66],
                   'X': [78, 85, 96, 80, 86],
                   'Y': [84, 94, 89, 83, 86],
                   'Z': [86, 97, 96, 72, 83]})

# Print the original DataFrame
print("Original DataFrame:")
print(df)

# Reverse column order
reversed_columns = df.iloc[:, ::-1]
print("Reverse column order:")
print(reversed_columns)

# Reverse row order
reversed_rows = df.iloc[::-1]
print("Reverse row order:")
print(reversed_rows)

# Reverse row order and reset index
reversed_rows_reset_index = df.iloc[::-1].reset_index(drop=True)
print("Reverse row order and reset index:")
print(reversed_rows_reset_index)


Original DataFrame:
    W   X   Y   Z
0  68  78  84  86
1  75  85  94  97
2  86  96  89  96
3  80  80  83  72
4  66  86  86  83
Reverse column order:
    Z   Y   X   W
0  86  84  78  68
1  97  94  85  75
2  96  89  96  86
3  72  83  80  80
4  83  86  86  66
Reverse row order:
    W   X   Y   Z
4  66  86  86  83
3  80  80  83  72
2  86  96  89  96
1  75  85  94  97
0  68  78  84  86
Reverse row order and reset index:
    W   X   Y   Z
0  66  86  86  83
1  80  80  83  72
2  86  96  89  96
3  75  85  94  97
4  68  78  84  86


# 66. Write a Pandas program to select columns by data type of a given DataFrame.
Sample Output:
Original DataFrame
name date_of_birth age
0 Alberto Franco 17/05/2002 18.5
1 Gino Mcneill 16/02/1999 21.2
2 Ryan Parkes 25/09/1998 22.5
3 Eesha Hinton 11/05/2002 22.0
4 Syed Wharton 15/09/1997 23.0
Select numerical columns
age
0 18.5
1 21.2
2 22.5
3 22.0
4 23.0
Select string columns
name date_of_birth
0 Alberto Franco 17/05/2002
1 Gino Mcneill 16/02/1999
2 Ryan Parkes 25/09/1998
3 Eesha Hinton 11/05/2002
4 Syed Wharton 15/09/1997

In [20]:
import pandas as pd

# Create the original DataFrame
df = pd.DataFrame({'name': ['Alberto Franco', 'Gino Mcneill', 'Ryan Parkes', 'Eesha Hinton', 'Syed Wharton'],
                   'date_of_birth': ['17/05/2002', '16/02/1999', '25/09/1998', '11/05/2002', '15/09/1997'],
                   'age': [18.5, 21.2, 22.5, 22.0, 23.0]})

# Print the original DataFrame
print("Original DataFrame:")
print(df)

# Select numerical columns
numerical_columns = df.select_dtypes(include=[float, int])
print("Select numerical columns:")
print(numerical_columns)

# Select string columns
string_columns = df.select_dtypes(include=[object])
print("Select string columns:")
print(string_columns)


Original DataFrame:
             name date_of_birth   age
0  Alberto Franco    17/05/2002  18.5
1    Gino Mcneill    16/02/1999  21.2
2     Ryan Parkes    25/09/1998  22.5
3    Eesha Hinton    11/05/2002  22.0
4    Syed Wharton    15/09/1997  23.0
Select numerical columns:
    age
0  18.5
1  21.2
2  22.5
3  22.0
4  23.0
Select string columns:
             name date_of_birth
0  Alberto Franco    17/05/2002
1    Gino Mcneill    16/02/1999
2     Ryan Parkes    25/09/1998
3    Eesha Hinton    11/05/2002
4    Syed Wharton    15/09/1997


# 67. Write a Pandas program to split a given DataFrame into two random subsets.
Sample Output:
Original Dataframe and shape:
name date_of_birth age
0 Alberto Franco 17/05/2002 18
1 Gino Mcneill 16/02/1999 21
2 Ryan Parkes 25/09/1998 22
3 Eesha Hinton 11/05/2002 22
4 Syed Wharton 15/09/1997 23
(5, 3)
Subset-1 and shape:
name date_of_birth age
1 Gino Mcneill 16/02/1999 21
4 Syed Wharton 15/09/1997 23
2 Ryan Parkes 25/09/1998 22
(3, 3)
Subset-2 and shape:
name date_of_birth age
0 Alberto Franco 17/05/2002 18
3 Eesha Hinton 11/05/2002 22
(2, 3)

In [21]:
import pandas as pd

# Create the original DataFrame
df = pd.DataFrame({'name': ['Alberto Franco', 'Gino Mcneill', 'Ryan Parkes', 'Eesha Hinton', 'Syed Wharton'],
                   'date_of_birth': ['17/05/2002', '16/02/1999', '25/09/1998', '11/05/2002', '15/09/1997'],
                   'age': [18, 21, 22, 22, 23]})

# Print the original DataFrame and its shape
print("Original DataFrame and shape:")
print(df)
print(df.shape)

# Split the DataFrame into two random subsets
subset1 = df.sample(frac=0.6, random_state=1)  # First subset with 60% of the original data
subset2 = df[~df.index.isin(subset1.index)]  # Second subset with the remaining data

# Print the first subset and its shape
print("Subset-1 and shape:")
print(subset1)
print(subset1.shape)

# Print the second subset and its shape
print("Subset-2 and shape:")
print(subset2)
print(subset2.shape)


Original DataFrame and shape:
             name date_of_birth  age
0  Alberto Franco    17/05/2002   18
1    Gino Mcneill    16/02/1999   21
2     Ryan Parkes    25/09/1998   22
3    Eesha Hinton    11/05/2002   22
4    Syed Wharton    15/09/1997   23
(5, 3)
Subset-1 and shape:
           name date_of_birth  age
2   Ryan Parkes    25/09/1998   22
1  Gino Mcneill    16/02/1999   21
4  Syed Wharton    15/09/1997   23
(3, 3)
Subset-2 and shape:
             name date_of_birth  age
0  Alberto Franco    17/05/2002   18
3    Eesha Hinton    11/05/2002   22
(2, 3)


# 68. Write a Pandas program to rename all columns with the same pattern of a given DataFrame.
Sample Output:
Original DataFrame
Name Date_Of_Birth Age
0 Alberto Franco 17/05/2002 18.5
1 Gino Mcneill 16/02/1999 21.2
2 Ryan Parkes 25/09/1998 22.5
3 Eesha Hinton 11/05/2002 22.0
4 Syed Wharton 15/09/1997 23.0
Remove trailing (at the end) whitesapce and convert to lowercase of the columns name
name date_of_birth age
0 Alberto Franco 17/05/2002 18.5
1 Gino Mcneill 16/02/1999 21.2
2 Ryan Parkes 25/09/1998 22.5
3 Eesha Hinton 11/05/2002 22.0
4 Syed Wharton 15/09/1997 23.0

In [22]:
import pandas as pd

# Create the original DataFrame
data = {'Name': ['Alberto Franco', 'Gino Mcneill', 'Ryan Parkes', 'Eesha Hinton', 'Syed Wharton'],
        'Date_Of_Birth': ['17/05/2002', '16/02/1999', '25/09/1998', '11/05/2002', '15/09/1997'],
        'Age': [18.5, 21.2, 22.5, 22.0, 23.0]}
df = pd.DataFrame(data)

# Rename columns based on the pattern
df = df.rename(columns=lambda x: x.strip().lower())

# Print the updated DataFrame
print(df)


             name date_of_birth   age
0  Alberto Franco    17/05/2002  18.5
1    Gino Mcneill    16/02/1999  21.2
2     Ryan Parkes    25/09/1998  22.5
3    Eesha Hinton    11/05/2002  22.0
4    Syed Wharton    15/09/1997  23.0


# 69. Write a Pandas program to merge datasets and check uniqueness.
Sample Output:
Original DataFrame:
Name Date_Of_Birth Age
0 Alberto Franco 17/05/2002 18.5
1 Gino Mcneill 16/02/1999 21.2
2 Ryan Parkes 25/09/1998 22.5
3 Eesha Hinton 11/05/2002 22.0
4 Syed Wharton 15/09/1997 23.0
New DataFrames:
Name Date_Of_Birth Age
2 Ryan Parkes 25/09/1998 22.5
3 Eesha Hinton 11/05/2002 22.0
4 Syed Wharton 15/09/1997 23.0
Name Date_Of_Birth Age
0 Alberto Franco 17/05/2002 18.5
1 Gino Mcneill 16/02/1999 21.2
3 Eesha Hinton 11/05/2002 22.0
4 Syed Wharton 15/09/1997 23.0
"one_to_one": check if merge keys are unique in both left and right datasets:"
Name Date_Of_Birth Age
0 Eesha Hinton 11/05/2002 22.0
1 Syed Wharton 15/09/1997 23.0
"one_to_many" or "1:m": check if merge keys are unique in left dataset:
Name Date_Of_Birth Age
0 Eesha Hinton 11/05/2002 22.0
1 Syed Wharton 15/09/1997 23.0
"any_to_one" or "m:1": check if merge keys are unique in right dataset:
Name Date_Of_Birth Age
0 Eesha Hinton 11/05/2002 22.0
1 Syed Wharton 15/09/1997 23.0

In [23]:
import pandas as pd

# Create the original DataFrame
df = pd.DataFrame({'Name': ['Alberto Franco', 'Gino Mcneill', 'Ryan Parkes', 'Eesha Hinton', 'Syed Wharton'],
                   'Date_Of_Birth': ['17/05/2002', '16/02/1999', '25/09/1998', '11/05/2002', '15/09/1997'],
                   'Age': [18.5, 21.2, 22.5, 22.0, 23.0]})

# Create the new DataFrames to merge
new_df1 = pd.DataFrame({'Name': ['Ryan Parkes', 'Eesha Hinton', 'Syed Wharton'],
                        'Date_Of_Birth': ['25/09/1998', '11/05/2002', '15/09/1997'],
                        'Age': [22.5, 22.0, 23.0]})

new_df2 = pd.DataFrame({'Name': ['Alberto Franco', 'Gino Mcneill', 'Eesha Hinton', 'Syed Wharton'],
                        'Date_Of_Birth': ['17/05/2002', '16/02/1999', '11/05/2002', '15/09/1997'],
                        'Age': [18.5, 21.2, 22.0, 23.0]})

# Print the original DataFrame
print("Original DataFrame:")
print(df)

# Print the new DataFrames
print("New DataFrames:")
print(new_df1)
print(new_df2)

# Perform the merge and check uniqueness
one_to_one_merge = pd.merge(df, new_df1, on=['Name', 'Date_Of_Birth', 'Age'], how='inner')
one_to_many_merge = pd.merge(df, new_df2, on=['Name'], how='inner')
any_to_one_merge = pd.merge(df, new_df2, on=['Name'], how='right')

# Print the merge results
print('"one_to_one": check if merge keys are unique in both left and right datasets:')
print(one_to_one_merge)
print('"one_to_many" or "1:m": check if merge keys are unique in left dataset:')
print(one_to_many_merge)
print('"any_to_one" or "m:1": check if merge keys are unique in right dataset:')
print(any_to_one_merge)


Original DataFrame:
             Name Date_Of_Birth   Age
0  Alberto Franco    17/05/2002  18.5
1    Gino Mcneill    16/02/1999  21.2
2     Ryan Parkes    25/09/1998  22.5
3    Eesha Hinton    11/05/2002  22.0
4    Syed Wharton    15/09/1997  23.0
New DataFrames:
           Name Date_Of_Birth   Age
0   Ryan Parkes    25/09/1998  22.5
1  Eesha Hinton    11/05/2002  22.0
2  Syed Wharton    15/09/1997  23.0
             Name Date_Of_Birth   Age
0  Alberto Franco    17/05/2002  18.5
1    Gino Mcneill    16/02/1999  21.2
2    Eesha Hinton    11/05/2002  22.0
3    Syed Wharton    15/09/1997  23.0
"one_to_one": check if merge keys are unique in both left and right datasets:
           Name Date_Of_Birth   Age
0   Ryan Parkes    25/09/1998  22.5
1  Eesha Hinton    11/05/2002  22.0
2  Syed Wharton    15/09/1997  23.0
"one_to_many" or "1:m": check if merge keys are unique in left dataset:
             Name Date_Of_Birth_x  Age_x Date_Of_Birth_y  Age_y
0  Alberto Franco      17/05/2002   18.5    

# 70. Write a Pandas program to convert continuous values of a column in a given DataFrame to categorical.
Input:
{ 'Name': ['Alberto Franco','Gino Mcneill','Ryan Parkes', 'Eesha Hinton', 'Syed Wharton'],
'Age': [18, 22, 40, 50, 80, 5] }
Output:
Age group:
0 kids
1 adult
2 elderly
3 adult
4 elderly
5 kids
Name: age_groups, dtype: category
Categories (3, object): [kids < adult < elderly]

In [24]:
import pandas as pd

# Create the input DataFrame
data = {'Name': ['Alberto Franco', 'Gino Mcneill', 'Ryan Parkes', 'Eesha Hinton', 'Syed Wharton'],
        'Age': [18, 22, 40, 50, 80, 5]}
df = pd.DataFrame(data)

# Define the age group boundaries and labels
age_bins = [0, 18, 60, float('inf')]
age_labels = ['kids', 'adult', 'elderly']

# Convert the 'Age' column to categorical
df['age_groups'] = pd.cut(df['Age'], bins=age_bins, labels=age_labels)

# Print the output
print("Age group:")
print(df['age_groups'])



ValueError: All arrays must be of the same length

# 71. Write a Pandas program to display memory usage of a given DataFrame and every column of the DataFrame.
Sample Output:
Original DataFrame:
Name Date_Of_Birth Age
0 Alberto Franco 17/05/2002 18.5
1 Gino Mcneill 16/02/1999 21.2
2 Ryan Parkes 25/09/1998 22.5
3 Eesha Hinton 11/05/2002 22.0
4 Syed Wharton 15/09/1997 23.0
Global usage of memory of the DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 3 columns):
Name 5 non-null object
Date_Of_Birth 5 non-null object
Age 5 non-null float64
dtypes: float64(1), object(2)
memory usage: 801.0 bytes
None
The usage of memory of every column of the said DataFrame:
Index 80
Name 346
Date_Of_Birth 335
Age 40
dtype: int64

In [ ]:
import pandas as pd

# Create the original DataFrame
data = {'Name': ['Alberto Franco', 'Gino Mcneill', 'Ryan Parkes', 'Eesha Hinton', 'Syed Wharton'],
        'Date_Of_Birth': ['17/05/2002', '16/02/1999', '25/09/1998', '11/05/2002', '15/09/1997'],
        'Age': [18.5, 21.2, 22.5, 22.0, 23.0]}
df = pd.DataFrame(data)

# Print the original DataFrame
print("Original DataFrame:")
print(df)

# Display memory usage of the DataFrame
print("Global usage of memory of the DataFrame:")
print(df.info())

# Display memory usage of every column of the DataFrame
print("The usage of memory of every column of the DataFrame:")
print(df.memory_usage())


# 72. Write a Pandas program to combine many given series to create a DataFrame.
Sample Output:
Original Series:
0 php
1 python
2 java
3 c#
4 c++
dtype: object
0 1
1 2
2 3
3 4
4 5
dtype: int64
Combine above series to a dataframe:
index 0
0 1 python
1 2 java
2 3 c#
3 4 c++
4 5 NaN
Using pandas concat:
0 1
0 php 1
1 python 2
2 java 3
3 c# 4
4 c++ 5
Using pandas DataFrame with a dictionary, gives a specific name to the columns:
col1 col2
0 php 1
1 python 2
2 java 3
3 c# 4
4 c++ 5

In [ ]:
import pandas as pd

# Create the original series
series1 = pd.Series(['php', 'python', 'java', 'c#', 'c++'])
series2 = pd.Series([1, 2, 3, 4, 5])

# Print the original series
print("Original Series:")
print(series1)
print(series2)

# Combine the series using pd.concat()
concatenated_df = pd.concat([series1, series2], axis=1)
print("Combine above series to a DataFrame using pd.concat():")
print(concatenated_df)

# Combine the series using DataFrame with a dictionary
dict_df = pd.DataFrame({'col1': series1, 'col2': series2})
print("Combine above series to a DataFrame using DataFrame with a dictionary:")
print(dict_df)


# 73. Write a Pandas program to create DataFrames that contains random values, contains missing values, contains datetime values and contains mixed values.
Sample Output:
DataFrame: Contains random values:
A B C D
Dog2w4Dv4l 0.591058 1.883454 -1.608613 -0.502516
kV7mfdFcF9 0.629642 -0.474377 0.567357 1.658445
.......
DataFrame: Contains missing values:
A B C D
i6i6Xn9l9c -0.299335 0.410871 -0.431840 -0.302177
OGo5KNNYNJ -0.174594 -1.366146 0.435063 -2.779446
u0mG9q1L7C 1.019094 -0.061077 -1.138138 -0.218460
RNJGqpci4o -0.380815 0.189970 -2.148521 -1.163589
vXIcxItZ1D NaN -0.079448 0.604777 0.065290
........
DataFrame: Contains datetime values:
A B C D
2000-01-03 0.665402 0.860808 -0.180986 -0.970889
2000-01-04 -1.511533 0.487539 -0.710355 -0.807816
2000-01-05 -0.773294 0.197918 -1.214035 1.049529
2000-01-06 -1.074894 1.774147 -0.620025 0.740779
.......
DataFrame: Contains mixed values:
A B C D
0 0.0 0.0 foo1 2009-01-01
1 1.0 1.0 foo2 2009-01-02
2 2.0 0.0 foo3 2009-01-05
3 3.0 1.0 foo4 2009-01-06
4 4.0 0.0 foo5 2009-01-07

In [ ]:
import pandas as pd
import numpy as np

# Create DataFrame with random values
df_random = pd.DataFrame(np.random.randn(5, 4), columns=['A', 'B', 'C', 'D'])
print("DataFrame: Contains random values:")
print(df_random)

# Create DataFrame with missing values
df_missing = pd.DataFrame({'A': [-0.299335, -0.174594, 1.019094, -0.380815, np.nan],
                           'B': [0.410871, -1.366146, -0.061077, 0.189970, -0.079448],
                           'C': [-0.431840, 0.435063, -1.138138, -2.148521, 0.604777],
                           'D': [-0.302177, -2.779446, -0.218460, -1.163589, 0.065290]})
print("DataFrame: Contains missing values:")
print(df_missing)

# Create DataFrame with datetime values
df_datetime = pd.DataFrame({'A': pd.to_datetime(['2000-01-03', '2000-01-04', '2000-01-05', '2000-01-06']),
                            'B': [0.665402, -1.511533, -0.773294, -1.074894],
                            'C': [0.860808, 0.487539, 0.197918, 1.774147],
                            'D': [-0.180986, -0.710355, -1.214035, -0.620025]})
print("DataFrame: Contains datetime values:")
print(df_datetime)

# Create DataFrame with mixed values
df_mixed = pd.DataFrame({'A': [0.0, 1.0, 2.0, 3.0, 4.0],
                         'B': [0.0, 1.0, 0.0, 1.0, 0.0],
                         'C': ['foo1', 'foo2', 'foo3', 'foo4', 'foo5'],
                         'D': pd.to_datetime(['2009-01-01', '2009-01-02', '2009-01-05', '2009-01-06', '2009-01-07'])})
print("DataFrame: Contains mixed values:")
print(df_mixed)


# 74. Write a Pandas program to fill missing values in time series data.
From Wikipedia , in the mathematical field of numerical analysis, interpolation is a type of estimation, a method of constructing new data points within the range of a discrete set of known data points.
Sample Output:
Original DataFrame:
c1 c2
2000-01-03 120.0 7.0
2000-01-04 130.0 NaN
2000-01-05 140.0 10.0
2000-01-06 150.0 NaN
2000-01-07 NaN 5.5
2000-01-10 170.0 16.5
DataFrame after interpolate:
c1 c2
2000-01-03 120.0 7.00
2000-01-04 130.0 8.50
2000-01-05 140.0 10.00
2000-01-06 150.0 7.75
2000-01-07 160.0 5.50
2000-01-10 170.0 16.50

In [ ]:
import pandas as pd

# Create the original DataFrame
df = pd.DataFrame({'c1': [120.0, 130.0, 140.0, 150.0, None, 170.0],
                   'c2': [7.0, None, 10.0, None, 5.5, 16.5]},
                  index=pd.to_datetime(['2000-01-03', '2000-01-04', '2000-01-05',
                                       '2000-01-06', '2000-01-07', '2000-01-10']))

# Fill missing values using interpolation
df_interpolated = df.interpolate()

# Print the resulting DataFrame
print("DataFrame after interpolate:")
print(df_interpolated)


# 75. Write a Pandas program to use a local variable within a query.
Sample Output:
Original DataFrame
W X Y Z
0 68 78 84 86
1 75 85 94 97
2 86 96 89 96
3 80 80 83 72
4 66 86 86 83
Values which are less than maximum value of 'W' column
W X Y Z
0 68 78 84 86
1 75 85 94 97
3 80 80 83 72
4 66 86 86 83

In [ ]:
import pandas as pd

# Create the original DataFrame
df = pd.DataFrame({'W': [68, 75, 86, 80, 66],
                   'X': [78, 85, 96, 80, 86],
                   'Y': [84, 94, 89, 83, 86],
                   'Z': [86, 97, 96, 72, 83]})

# Define the local variable
max_value_W = df['W'].max()

# Use the local variable within the query
result = df.query("W < @max_value_W")

# Print the resulting DataFrame
print("Values which are less than maximum value of 'W' column:")
print(result)


# 76. Write a Pandas program to clean object column with mixed data of a given DataFrame using regular expression.
Sample Output:
Original dataframe:
agent purchase
0 a001 4500
1 a002 7500
2 a003 $3000.25
3 a003 $1250.35
4 a004 9000.00
Data Types:
0 <class 'float'>
1 <class 'float'>
2 <class 'str'>
3 <class 'str'>
4 <class 'str'>
Name: purchase, dtype: object
New Data Types:
0 <class 'float'>
1 <class 'float'>
2 <class 'float'>
3 <class 'float'>
4 <class 'float'>
Name: purchase, dtype: object

In [ ]:
import pandas as pd

# Create the original DataFrame
data = {'agent': ['a001', 'a002', 'a003', 'a003', 'a004'],
        'purchase': ['4500', '7500', '$3000.25', '$1250.35', '9000.00']}
df = pd.DataFrame(data)

# Print the original DataFrame
print("Original dataframe:")
print(df)

# Clean the 'purchase' column using regular expressions
df['purchase'] = df['purchase'].str.replace(r'[^\d.]', '', regex=True).astype(float)

# Print the data types of the 'purchase' column
print("Data Types:")
print(df['purchase'].apply(type))

# Print the new data types of the 'purchase' column
print("New Data Types:")
print(df['purchase'].apply(type).astype(str))


# 77. Write a Pandas program to get the numeric representation of an array by identifying distinct values of a given column of a dataframe.
Sample Output:
Original DataFrame:
Name Date_Of_Birth Age
0 Alberto Franco 17/05/2002 18.5
1 Gino Mcneill 16/02/1999 21.2
2 Ryan Parkes 25/09/1998 22.5
3 Eesha Hinton 11/05/2002 22.0
4 Gino Mcneill 15/09/1997 23.0
Numeric representation of an array by identifying distinct values:
[0 1 2 3 1]
Index(['Alberto Franco', 'Gino Mcneill', 'Ryan Parkes', 'Eesha Hinton'], dtype='object')

In [ ]:
import pandas as pd
import numpy as np

# Create the original DataFrame
data = {'Name': ['Alberto Franco', 'Gino Mcneill', 'Ryan Parkes', 'Eesha Hinton', 'Gino Mcneill'],
        'Date_Of_Birth': ['17/05/2002', '16/02/1999', '25/09/1998', '11/05/2002', '15/09/1997'],
        'Age': [18.5, 21.2, 22.5, 22.0, 23.0]}
df = pd.DataFrame(data)

# Print the original DataFrame
print("Original DataFrame:")
print(df)

# Get the numeric representation of the column
numeric_representation = pd.factorize(df['Name'])[0]

# Print the numeric representation
print("Numeric representation of an array by identifying distinct values:")
print(numeric_representation)
print("Index:")
print(pd.factorize(df['Name'])[1])


# 78. Write a Pandas program to replace the current value in a dataframe column based on last largest value. If the current value is less than last largest value replaces the value with 0.
Test data:
rnum
0 23
1 21
2 27
3 22
...
10 34
11 19
12 31
13 32
14 19
Sample Output:
Original DataFrame:
rnum
0 23
1 21
2 27
3 22
...
10 34
11 19
12 31
13 32
14 19
Replace current value in a dataframe column based on last largest value:
rnum
0 23
1 0
2 27
3 0
...
10 34
11 0
12 0
13 0
14 0


In [ ]:
import pandas as pd

# Create the original DataFrame
data = {'rnum': [23, 21, 27, 22, 34, 19, 31, 32, 19]}
df = pd.DataFrame(data)

# Print the original DataFrame
print("Original DataFrame:")
print(df)

# Replace current values based on last largest value
last_largest = 0
for i, value in enumerate(df['rnum']):
    if value < last_largest:
        df.at[i, 'rnum'] = 0
    else:
        last_largest = value

# Print the updated DataFrame
print("Replace current value in a DataFrame column based on last largest value:")
print(df)


# 79. Write a Pandas program to create a DataFrame from the clipboard (data from an Excel spreadsheet or a Google Sheet).
Sample Excel Data:
Python Exercises: Sample Excel data.
Sample Output:
Data from clipboard to DataFrame:
1 2 3 4
0 2 3 4 5
1 4 5 1 0
2 2 3 7 8

In [27]:
import pandas as pd

# Read data from clipboard to DataFrame
df = pd.read_clipboard(sep='\s')

# Print the DataFrame
print("Data from clipboard to DataFrame:")
print(df)


ParserError: Expected 4 fields in line 6, saw 5. Error could possibly be due to quotes being ignored when a multi-char delimiter is used.

# 80. Write a Pandas program to check for inequality of two given DataFrames.
Sample Output:
Original DataFrames:
W X Y Z
0 68.0 78.0 84 86
1 75.0 85.0 94 97
2 86.0 NaN 89 96
3 80.0 80.0 83 72
4 NaN 86.0 86 83
W X Y Z
0 78.0 78 84 86
1 75.0 85 84 97
2 86.0 96 89 96
3 80.0 80 83 72
4 NaN 76 86 83
Check for inequality of the said dataframes:
W X Y Z
0 True False False False
1 False False True False
2 False True False False
3 False False False False
4 True True False False

In [ ]:
import pandas as pd
import numpy as np

# Original DataFrames
df1 = pd.DataFrame({'W': [68.0, 75.0, 86.0, 80.0, np.nan],
                    'X': [78.0, 85.0, np.nan, 80.0, 86.0],
                    'Y': [84, 94, 89, 83, 86],
                    'Z': [86, 97, 96, 72, 83]})

df2 = pd.DataFrame({'W': [78.0, 75.0, 86.0, 80.0, np.nan],
                    'X': [78, 85, 96, 80, 76],
                    'Y': [84, 84, 89, 83, 86],
                    'Z': [86, 97, 96, 72, 83]})

# Check for inequality of the two DataFrames
inequality = df1 != df2

# Print the inequality DataFrame
print("Check for inequality of the two DataFrames:")
print(inequality)


# 81. Write a Pandas program to get lowest n records within each group of a given DataFrame.
Sample Output:
Original DataFrame
col1 col2 col3
0 1 4 7
1 2 5 5
2 3 6 8
3 4 9 12
4 7 5 1
5 11 0 11
Lowest n records within each group of a DataFrame:
col1 col2 col3
0 1 4 7
1 2 5 5
2 3 6 8
col1 col2 col3
5 11 0 11
0 1 4 7
1 2 5 5
col1 col2 col3
4 7 5 1
1 2 5 5
0 1 4 7

In [ ]:
import pandas as pd

# Original DataFrame
df = pd.DataFrame({'col1': [1, 2, 3, 4, 7, 11],
                   'col2': [4, 5, 6, 9, 5, 0],
                   'col3': [7, 5, 8, 12, 1, 11]})

# Function to get the lowest n records within each group
def get_lowest_n_records(group, n):
    return group.nsmallest(n, 'col1')

# Group by 'col1' and get the lowest 2 records within each group
lowest_n_records = df.groupby('col1').apply(get_lowest_n_records, n=2).reset_index(drop=True)

# Print the lowest n records within each group
print("Lowest 2 records within each group of the DataFrame:")
print(lowest_n_records)
